### PR TITLE
feat(JSON-RPC): make transaction hashes optional in get_block

### PR DESCRIPTION
**Pull request by aesedepece**
This speeds up synchronization times by 10x.

### CODE REVIEW

**Comment by tmpolaczyk** [comment]
We should consider using a graphql api, since that's their main selling point: no unused data on query responses.

File: node/src/actors/json_rpc/json_rpc_methods.rs [hunk 1]
```diff
@@ -534,13 +534,17 @@ pub fn get_block_chain(
...
 /* test
-{"jsonrpc":"2.0","id":1,"method":"getBlock","params":["c0002c6b25615c0f71069f159dffddf8a0b3e529efb054402f0649e969715bdb"]}
-{"jsonrpc":"2.0","id":1,"method":"getBlock","params":[{"SHA256":[255,198,135,145,253,40,66,175,226,220,119,243,233,210,25,119,171,217,215,188,185,190,93,116,164,234,217,67,30,102,205,46]}]}
+{"jsonrpc":"2.0","id":1,"method":"getBlock","params":["c0002c6b25615c0f71069f159dffddf8a0b3e529efb054402f0649e969715bdb", false]}
+{"jsonrpc":"2.0","id":1,"method":"getBlock","params":[{"SHA256":[255,198,135,145,253,40,66,175,226,220,119,243,233,210,25,119,171,217,215,188,185,190,93,116,164,234,217,67,30,102,205,46, false]}]}
 */
-pub fn get_block(hash: Result<(Hash,), jsonrpc_core::Error>) -> JsonRpcResultAsync {
-    let hash = match hash {
-        Ok(x) => x.0,
+pub fn get_block(args: Result<(Hash, bool), jsonrpc_core::Error>) -> JsonRpcResultAsync {
```
Comment (tmpolaczyk) [inline node/src/actors/json_rpc/json_rpc_methods.rs:545]: This is a breaking change, but there is a way to make it non-breaking and default the bool to true to keep the old behavior. We need to pass the args as a json value and do the parsing manually.

```suggestion
pub fn get_block(args: Value) -> JsonRpcResultAsync {
```

Comment (aesedepece) [reply node/src/actors/json_rpc/json_rpc_methods.rs]: Done :wink:

File: node/src/actors/json_rpc/json_rpc_methods.rs [hunk 2]
```diff
@@ -534,64 +536,103 @@ pub fn get_block_chain(
...
-    let hash = match hash {
-        Ok(x) => x.0,
-        Err(e) => return Box::new(futures::failed(e)),
+pub fn get_block(params: Params) -> JsonRpcResultAsync {
+    let (block_hash, include_txns_hashes): (Hash, bool);
+
+    // Handle parameters as an array with a first obligatory hash field plus an optional bool field
+    if let Params::Array(params) = params {
+        if let Some(Value::String(hash)) = params.get(0) {
+            block_hash = hash.parse().unwrap()
```
Comment (tmpolaczyk) [inline node/src/actors/json_rpc/json_rpc_methods.rs:553]: This checks that `hash` is a string, but it doesn't check if that string is a valid `Hash`. Try using a match like the one that was here before:

```rust
   let hash = match hash.parse() {
        Ok(x) => x,
        Err(e) => return Box::new(futures::failed(e)),
   };
```

**Comment by tmpolaczyk** [comment]
I made a few tests to check that the params are deserialized correctly, but unfortunately we cannot check if the flag is read as true or false:

```rust
    #[test]
    fn get_block_one_param() {
        use witnet_data_structures::chain::*;

        // 1-element tuple is serialized as 1-element array: ["0000..."]
        let params = ("0000000000000000000000000000000000000000000000000000000000000000",);
        let s = serde_json::to_string(&params).unwrap();
        let msg = format!(
            r#"{{"jsonrpc":"2.0","method":"getBlock","params":{},"id":1}}"#,
            s
        );

        // Expected result: true
        let expected = r#"{"jsonrpc":"2.0","error":{"code":-32603,"message":"MailboxError(Mailbox has closed)"},"id":1}"#.to_string();
        let subscriptions = Subscriptions::default();
        let (transport_sender, _transport_receiver) = mpsc::channel(0);
        let meta = Arc::new(Session::new(transport_sender));
        let io = jsonrpc_io_handler(subscriptions, true);
        let response = io.handle_request_sync(&msg, meta);
        assert_eq!(response, Some(expected));
    }

    #[test]
    fn get_block_two_param() {
        use witnet_data_structures::chain::*;

        let params = ("0000000000000000000000000000000000000000000000000000000000000000", true);
        let s = serde_json::to_string(&params).unwrap();
        let msg = format!(
            r#"{{"jsonrpc":"2.0","method":"getBlock","params":{},"id":1}}"#,
            s
        );

        // Expected result: true
        let expected = r#"{"jsonrpc":"2.0","error":{"code":-32603,"message":"MailboxError(Mailbox has closed)"},"id":1}"#.to_string();
        let subscriptions = Subscriptions::default();
        let (transport_sender, _transport_receiver) = mpsc::channel(0);
        let meta = Arc::new(Session::new(transport_sender));
        let io = jsonrpc_io_handler(subscriptions, true);
        let response = io.handle_request_sync(&msg, meta);
        assert_eq!(response, Some(expected));
    }
```

**Comment by aesedepece** [comment]
> I made a few tests to check that the params are deserialized correctly, but unfortunately we cannot check if the flag is read as true or false:
> 
> ```rust
>     #[test]
>     fn get_block_one_param() {
>         use witnet_data_structures::chain::*;
> 
>         // 1-element tuple is serialized as 1-element array: ["0000..."]
>         let params = ("0000000000000000000000000000000000000000000000000000000000000000",);
>         let s = serde_json::to_string(&params).unwrap();
>         let msg = format!(
>             r#"{{"jsonrpc":"2.0","method":"getBlock","params":{},"id":1}}"#,
>             s
>         );
> 
>         // Expected result: true
>         let expected = r#"{"jsonrpc":"2.0","error":{"code":-32603,"message":"MailboxError(Mailbox has closed)"},"id":1}"#.to_string();
>         let subscriptions = Subscriptions::default();
>         let (transport_sender, _transport_receiver) = mpsc::channel(0);
>         let meta = Arc::new(Session::new(transport_sender));
>         let io = jsonrpc_io_handler(subscriptions, true);
>         let response = io.handle_request_sync(&msg, meta);
>         assert_eq!(response, Some(expected));
>     }
> 
>     #[test]
>     fn get_block_two_param() {
>         use witnet_data_structures::chain::*;
> 
>         let params = ("0000000000000000000000000000000000000000000000000000000000000000", true);
>         let s = serde_json::to_string(&params).unwrap();
>         let msg = format!(
>             r#"{{"jsonrpc":"2.0","method":"getBlock","params":{},"id":1}}"#,
>             s
>         );
> 
>         // Expected result: true
>         let expected = r#"{"jsonrpc":"2.0","error":{"code":-32603,"message":"MailboxError(Mailbox has closed)"},"id":1}"#.to_string();
>         let subscriptions = Subscriptions::default();
>         let (transport_sender, _transport_receiver) = mpsc::channel(0);
>         let meta = Arc::new(Session::new(transport_sender));
>         let io = jsonrpc_io_handler(subscriptions, true);
>         let response = io.handle_request_sync(&msg, meta);
>         assert_eq!(response, Some(expected));
>     }
> ```

Do you mean that it can't be tested because `true` is used as default? If so, we should still be able to test that, if present and set to `false`, it is read.

File: node/src/actors/json_rpc/json_rpc_methods.rs [hunk 3]
```diff
@@ -44,6 +40,12 @@ use crate::{
     config_mngr, signature_mngr,
 };
 
+//use std::str::FromStr;
```
Comment (girazoki) [inline node/src/actors/json_rpc/json_rpc_methods.rs:43]:
```suggestion

```

File: node/src/actors/json_rpc/json_rpc_methods.rs [hunk 4]
```diff
@@ -534,64 +536,103 @@ pub fn get_block_chain(
...
+        } else {
+            return Box::new(futures::future::Either::<FutureResult<_, _>, _>::B(
+                futures::failed(jsonrpc_core::Error::invalid_params(
+                    "First argument of `get_block` must have type `Hash`",
+                )),
+            ));
+        };
+
+        match params.get(1) {
+            None => include_txns_hashes = true,
```
Comment (girazoki) [inline node/src/actors/json_rpc/json_rpc_methods.rs:563]: Does this mean that by default we get the transaction hashes?

Comment (aesedepece) [reply node/src/actors/json_rpc/json_rpc_methods.rs:563]: Yes, so as not to break compatibility with any existing clients (e.g. block explorer)

**Comment by tmpolaczyk** [comment]
> @tmpolaczyk do you mean that it can't be tested because `true` is used as default? If so, we should still be able to test that, if present and set to `false`, it is read.

That test checks if the result is "MailboxError", so we don't know if the block with have the transactions hashes or not.

File: node/src/actors/json_rpc/json_rpc_methods.rs (hunk 4 above)
Comment (drcpu-github) [reply node/src/actors/json_rpc/json_rpc_methods.rs:563]: If it is an overall improvement to disable this by default, feel free to break my explorer. I'm probably the only one who uses that functionality. It's not that hard to modify, but just let me know before it goes live. 😄

File: node/src/actors/json_rpc/json_rpc_methods.rs (hunk 3 above)
Comment (aesedepece) [reply node/src/actors/json_rpc/json_rpc_methods.rs]: Resolved

File: node/src/actors/json_rpc/json_rpc_methods.rs [hunk 5]
```diff
@@ -534,64 +535,110 @@ pub fn get_block_chain(
...
 /// Get block by hash
+///
+/// - First argument is the hash of the block that we are querying.
+/// - Second argument is whether we want the response to contain a list of hashes of the
+///   transactions found in the block.  
 /* test
-{"jsonrpc":"2.0","id":1,"method":"getBlock","params":["c0002c6b25615c0f71069f159dffddf8a0b3e529efb054402f0649e969715bdb"]}
-{"jsonrpc":"2.0","id":1,"method":"getBlock","params":[{"SHA256":[255,198,135,145,253,40,66,175,226,220,119,243,233,210,25,119,171,217,215,188,185,190,93,116,164,234,217,67,30,102,205,46]}]}
+{"jsonrpc":"2.0","id":1,"method":"getBlock","params":["c0002c6b25615c0f71069f159dffddf8a0b3e529efb054402f0649e969715bdb", false]}
+{"jsonrpc":"2.0","id":1,"method":"getBlock","params":[{"SHA256":[255,198,135,145,253,40,66,175,226,220,119,243,233,210,25,119,171,217,215,188,185,190,93,116,164,234,217,67,30,102,205,46]}, false]}
```
Comment (tmpolaczyk) [inline node/src/actors/json_rpc/json_rpc_methods.rs:544]: This syntax has not been supported for a long time

```suggestion
```

**Review by tmpolaczyk — APPROVED** [review]

**Pull request merged**